tracing/span.go

```diff
@@ -0,0 +1,240 @@
+package tracing
+
+import (
+	"context"
+	"math/rand"
+	"time"
+)
+
+// SpanType enum.
+type SpanType int
+
+// SpanType values.
+const (
+	SpanTypeLocal SpanType = iota
+	SpanTypeClient
+	SpanTypeServer
+)
+
+// FlagMask values.
+//
+// Reference: https://github.com/reddit/baseplate.py/blob/1ca8488bcd42c8786e6a3db35b2a99517fd07a99/baseplate/observers/tracing.py#L60-L64
+const (
+	// When set, this trace passes all samplers.
+	FlagMaskDebug int64 = 1
+)
+
+const (
+	counterKeyPrefix   = "counter."
+	baseplateComponent = "baseplate"
+)
+
+// Span defines a tracing span.
+type Span struct {
+	tracer *Tracer
+
+	name     string
+	traceID  uint64
+	spanID   uint64
+	parentID uint64
+	sampled  bool
+
+	component string
+
+	start    time.Time
+	end      time.Time
+	spanType SpanType
+	flags    int64
+
+	counters map[string]float64
+	tags     map[string]interface{}
+}
+
+func newSpan(tracer *Tracer, spanType SpanType) *Span {
+	if tracer == nil {
+		tracer = &GlobalTracer
+	}
+	return &Span{
+		tracer:   tracer,
+		spanID:   rand.Uint64(),
+		start:    time.Now(),
+		spanType: spanType,
+
+		counters: make(map[string]float64),
+		tags: map[string]interface{}{
+			ZipkinBinaryAnnotationKeyComponent: baseplateComponent,
+		},
+	}
+}
+
+// ToZipkinSpan returns a ZipkinSpan with data copied from this span.
+//
+// If it's called before End is called,
+// Duration and some annotations might be incorrect/incomplete.
```

Comment:
I exported this function mainly because it's needed in `thrift_test.go` to inspect the span metadata is correct, and also I want to keep that file an external test.

In most cases I don't just export a function for test, and would rather just make the test not external. But here exporting this function is harmless, as the external caller cannot do anything with the returned type of `ZipkinSpan`, which makes the value of keeping the test external high.

Reply:
I added a comment on `Tracer.Record` to mention that in most cases `Span.End` should be enough and that function shouldn't be called directly.